### PR TITLE
[Merged by Bors] - feat: port Combinatorics.Young.YoungDiagram

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/Young/YoungDiagram.lean [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+/-- A Young diagram is a finite collection of cells on the `ℕ × ℕ` grid such that whenever
+a cell is present, so are all the ones above and to the left of it. Like matrices, an `(i, j)` cell
+is a cell in row `i` and column `j`, where rows are enumerated downward and columns rightward.
+
+Young diagrams are modeled as finite sets in `ℕ × ℕ` that are lower sets with respect to the
+standard order on products. -/
+@[ext]
+structure YoungDiagram where
+  /-- A finite set which represents a finite collection of cells on the `ℕ × ℕ` grid. -/
+  cells : Finset (ℕ × ℕ)
```
Comment (Komyyy) [inline Mathlib/Combinatorics/Young/YoungDiagram.lean:67]: My English is not so good, please correct my docs for projections.

File: Mathlib/Combinatorics/Young/YoungDiagram.lean [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+  /-- Cells are up-left justified. We represent it whether the `cells` is a lower set in `ℕ × ℕ` in
+  the product partial order. -/
+  IsLowerSet : IsLowerSet (cells : Set (ℕ × ℕ))
```
Comment (Komyyy) [inline Mathlib/Combinatorics/Young/YoungDiagram.lean:70]: My English is not so good, please correct my docs for projections.

File: Mathlib/Combinatorics/Young/YoungDiagram.lean [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+is a cell in row `i` and column `j`, where rows are enumerated downward and columns rightward.
+
+Young diagrams are modeled as finite sets in `ℕ × ℕ` that are lower sets with respect to the
+standard order on products. -/
+@[ext]
+structure YoungDiagram where
+  /-- A finite set which represents a finite collection of cells on the `ℕ × ℕ` grid. -/
+  cells : Finset (ℕ × ℕ)
+  /-- Cells are up-left justified. We represent it whether the `cells` is a lower set in `ℕ × ℕ` in
+  the product partial order. -/
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Young/YoungDiagram.lean:69]:
```suggestion
  /-- Cells are up-left justified, witnessed by the fact that `cells` is a lower set in `ℕ × ℕ`. -/
```

File: Mathlib/Combinatorics/Young/YoungDiagram.lean [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+  IsLowerSet : IsLowerSet (cells : Set (ℕ × ℕ))
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Young/YoungDiagram.lean:70]:
```suggestion
  isLowerSet : IsLowerSet (cells : Set (ℕ × ℕ))
```

File: Mathlib/Combinatorics/Young/YoungDiagram.lean [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+/-- The empty Young diagram is (⊥ : young_diagram). -/
+instance : OrderBot YoungDiagram
+    where
+  bot :=
+    { cells := ∅
+      IsLowerSet := by
+        intros a b _ h
+        simp only [Finset.coe_empty, Set.mem_empty_iff_false]
+        simp only [Finset.coe_empty, Set.mem_empty_iff_false] at h
+        }
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Young/YoungDiagram.lean:168]:
```suggestion
instance : OrderBot YoungDiagram where
  bot :=
    { cells := ∅
      IsLowerSet := by
        intros a b _ h
        simp only [Finset.coe_empty, Set.mem_empty_iff_false]
        simp only [Finset.coe_empty, Set.mem_empty_iff_false] at h }
```

File: Mathlib/Combinatorics/Young/YoungDiagram.lean [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+#align young_diagram.card YoungDiagram.card
+
+section Transpose
+
+/-- The `transpose` of a Young diagram is obtained by swapping i's with j's. -/
+def transpose (μ : YoungDiagram) : YoungDiagram
+    where
+  cells := (Equiv.prodComm _ _).finsetCongr μ.cells
+  IsLowerSet _ _ h :=
+    by
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Young/YoungDiagram.lean:215]:
```suggestion
def transpose (μ : YoungDiagram) : YoungDiagram where
  cells := (Equiv.prodComm _ _).finsetCongr μ.cells
  IsLowerSet _ _ h := by
```

File: Mathlib/Combinatorics/Young/YoungDiagram.lean [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+#align young_diagram.mem_row_iff YoungDiagram.mem_row_iff
+
+theorem mk_mem_row_iff {μ : YoungDiagram} {i j : ℕ} : (i, j) ∈ μ.row i ↔ (i, j) ∈ μ := by simp [row]
+#align young_diagram.mk_mem_row_iff YoungDiagram.mk_mem_row_iff
+
+protected theorem exists_not_mem_row (μ : YoungDiagram) (i : ℕ) : ∃ j, (i, j) ∉ μ := by
+  obtain ⟨j, hj⟩ :=
+    Infinite.exists_not_mem_finset
+      (μ.cells.preimage (Prod.mk i) fun _ _ _ _ h =>
+        by
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Young/YoungDiagram.lean:305]:
```suggestion
      (μ.cells.preimage (Prod.mk i) fun _ _ _ _ h => by
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**